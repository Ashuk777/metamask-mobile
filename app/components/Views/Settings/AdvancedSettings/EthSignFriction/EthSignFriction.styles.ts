// Third party dependencies.
import { StyleSheet } from 'react-native';
import { fontStyles } from '../../../../../styles/common';
import Device from '../../../../../util/device';

const breakPoint = Device.getDeviceHeight() < 700;

/**
 * Style sheet function for EthSignFriction component.
 *
 * @returns StyleSheet object.
 */

export default (colors: any) =>
  StyleSheet.create({
    frictionContainer: {
      flexDirection: 'column',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      padding: 16,
      width: '100%',
      alignSelf: 'center',
    },
    input: {
      width: '100%',
      ...fontStyles.normal,
      fontSize: 16,
      paddingTop: 2,
      color: colors.text.default,
    },
    warningIcon: {
      alignSelf: 'center',
      color: colors.error.default,
      marginTop: 2,
      marginEnd: 3,
      marginBottom: 4,
      marginStart: 3,
    },
    heading: {
      marginHorizontal: 6,
      color: colors.text.default,
      ...fontStyles.bold,
      fontSize: 20,
      textAlign: 'center',
      lineHeight: breakPoint ? 24 : 26,
    },
    red: {
      marginHorizontal: 24,
      color: colors.error.default,
    },
    warning: {
      ...fontStyles.normal,
      fontSize: 14,
      lineHeight: breakPoint ? 18 : 22,
      color: colors.text.default,
      marginTop: 24,
      marginBottom: 16,
      backgroundColor: colors.error.muted,
      borderRadius: 4,
      borderLeftWidth: 4,
      borderLeftColor: colors.error.default,
      display: 'flex',
      flexDirection: 'row',
      padding: 8,
    },
    warningText: {
      flex: 1,
    },
    bold: {
      ...fontStyles.bold,
    },
    understandCheckbox: {
      marginEnd: 8,
      height: 20,
      width: 20,
    },
    understandCheckText: {
      marginEnd: 16,
    },
    understandCheckboxView: {
      margin: 16,
      flexDirection: 'row',
      alignItems: 'stretch',
      justifyContent: 'flex-start',
      display: 'flex',
      fontSize: 14,
    },
    buttonsContainer: {
      flexDirection: 'row',
      display: 'flex',
    },
    cancelContainerStyle: {
      marginHorizontal: 8,
      flex: 1,
    },
    areYouSure: {
      width: '100%',
      padding: breakPoint ? 16 : 24,
      justifyContent: 'center',
      alignSelf: 'center',
    },
  });
