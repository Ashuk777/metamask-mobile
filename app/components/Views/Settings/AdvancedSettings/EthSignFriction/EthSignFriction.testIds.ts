/* eslint-disable import/prefer-default-export */
export const ETH_SIGN_FRICTION_TEXTFIELD_TEST_ID =
  'eth-sign-friction-textfield';
