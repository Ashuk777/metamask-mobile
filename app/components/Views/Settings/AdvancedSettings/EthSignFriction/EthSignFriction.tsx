// Third party dependencies.
import React, { useMemo, useRef, useState } from 'react';
import { View } from 'react-native';

// External dependencies.
import SheetBottom, {
  SheetBottomRef,
} from '../../../../../component-library/components/Sheet/SheetBottom';
import { strings } from '../../../../../../locales/i18n';
import Text from '../../../../Base/Text';
import { useTheme } from '../../../../../util/theme';
import Icon, {
  IconName,
  IconSize,
} from '../../../../../component-library/components/Icons/Icon';
import Engine from '../../../../../core/Engine';
import { tlc } from '../../../../../util/general';
import generateTestId from '../../../../../../wdio/utils/generateTestId';
import TextField from '../../../../../component-library/components/Form/TextField';

// Internal dependencies
import createStyles from './EthSignFriction.styles';
import { ETH_SIGN_FRICTION_TEXTFIELD_TEST_ID } from './EthSignFriction.testIds';
import Checkbox from '../../../../../component-library/components/Checkbox';
import Button, {
  ButtonVariants,
  ButtonWidthTypes,
} from '../../../../../component-library/components/Buttons/Button';

const EthSignFriction = () => {
  const { colors, themeAppearance } = useTheme();
  const styles = createStyles(colors);
  const sheetRef = useRef<SheetBottomRef>(null);
  const [understandCheckbox, setUnderstandCheckbox] = useState(false);
  const [firstFrictionPassed, setFirstFrictionPassed] = useState(false);
  const [approveText, setApproveText] = useState<string>('');

  const setText = (text: string) => {
    setApproveText(text);
  };

  const isApproveTextMatched = (text: string) =>
    tlc(text) === tlc(strings('app_settings.toggleEthSignModalFormValidation'));

  const toggleUnderstandCheckbox = () => {
    setUnderstandCheckbox(!understandCheckbox);
  };

  const onLearnMorePress = () => {
    //TODO navigate to support center
    //eslint-disable-next-line no-console
    console.log('learn more pressed');
  };

  const onCancelPress = () => {
    sheetRef.current?.hide();
  };

  const onPrimaryPress = () => {
    if (!firstFrictionPassed && understandCheckbox) {
      setFirstFrictionPassed(true);
    } else if (firstFrictionPassed && isApproveTextMatched(approveText)) {
      const { PreferencesController } = Engine.context;
      PreferencesController.setDisabledRpcMethodPreference('eth_sign', true);
      sheetRef.current?.hide();
    }
  };

  const isReadyToEnable = useMemo(
    () =>
      (understandCheckbox && !firstFrictionPassed) ||
      (firstFrictionPassed && isApproveTextMatched(approveText)),
    [understandCheckbox, firstFrictionPassed, approveText],
  );

  return (
    <SheetBottom ref={sheetRef}>
      <View style={styles.frictionContainer}>
        <Icon
          style={styles.warningIcon}
          color={colors.error.default}
          name={IconName.Danger}
          size={IconSize.Xl}
        />
        <Text style={styles.heading}>
          {strings('app_settings.toggleEthSignModalTitle')}
        </Text>
        <Text>
          <Text>{strings('app_settings.toggleEthSignModalDescription')} </Text>
          <Text
            primary
            noMargin
            link
            accessibilityRole={'link'}
            onPress={onLearnMorePress}
          >
            {strings('app_settings.toggleEthSignModalLearnMore')}
          </Text>
        </Text>
        <View style={styles.warning}>
          <Icon
            style={styles.warningIcon}
            color={colors.error.default}
            name={IconName.Danger}
            size={IconSize.Lg}
          />
          <Text style={styles.warningText}>
            <Text>{strings('app_settings.toggleEthSignModalBannerText')}</Text>
            <Text style={styles.bold}>
              {strings('app_settings.toggleEthSignModalBannerBoldText')}
            </Text>
          </Text>
        </View>
        {!firstFrictionPassed ? (
          <View style={[styles.understandCheckboxView]}>
            <Checkbox
              isChecked={understandCheckbox}
              style={styles.understandCheckbox}
              onPress={toggleUnderstandCheckbox}
              accessibilityRole={'checkbox'}
            />
            <Text
              onPress={() => toggleUnderstandCheckbox()}
              style={styles.understandCheckText}
            >
              {strings('app_settings.toggleEthSignModalCheckBox')}
            </Text>
          </View>
        ) : (
          <View style={styles.areYouSure}>
            <Text style={[styles.textConfirmField, styles.bold]}>
              {strings('app_settings.toggleEthSignModalFormLabel')}
            </Text>
            <TextField
              contextMenuHidden
              autoCompleteType={'off'}
              autoCorrect={false}
              isError={approveText.length > 0 && !isReadyToEnable}
              returnKeyType={'done'}
              onEndEditing={(e) => {
                setText(e.nativeEvent.text);
              }}
              autoCapitalize="none"
              onFocus={() => setApproveText('')}
              keyboardAppearance={themeAppearance}
              {...generateTestId(ETH_SIGN_FRICTION_TEXTFIELD_TEST_ID)}
            />
            {approveText.length > 0 && !isReadyToEnable && (
              <Text style={[styles.red, styles.textConfirmWarningText]}>
                {strings('app_settings.toggleEthSignModalFormError')}
              </Text>
            )}
          </View>
        )}
        <View style={styles.buttonsContainer}>
          <Button
            variant={ButtonVariants.Secondary}
            width={ButtonWidthTypes.Full}
            label={strings('navigation.cancel')}
            onPress={onCancelPress}
            style={styles.buttonStart}
            accessibilityRole={'button'}
          />

          <Button
            variant={ButtonVariants.Primary}
            width={ButtonWidthTypes.Full}
            isDanger={firstFrictionPassed}
            label={strings(
              firstFrictionPassed
                ? 'app_settings.toggleEthSignEnableButton'
                : 'app_settings.toggleEthSignContinueButton',
            )}
            onPress={onPrimaryPress}
            disabled={!isReadyToEnable}
            style={styles.buttonEnd}
            accessibilityRole={'button'}
          />
        </View>
      </View>
    </SheetBottom>
  );
};

export default EthSignFriction;
