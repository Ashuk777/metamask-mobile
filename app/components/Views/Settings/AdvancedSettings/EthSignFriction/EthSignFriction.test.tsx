import { shallow } from 'enzyme';
import React from 'react';
import EthSignFriction from './EthSignFriction';
import Engine from '../../../../../core/Engine';
import renderWithProvider from '../../../../../util/test/renderWithProvider';
import { fireEvent, screen } from '@testing-library/react-native';
import { strings } from '../../../../../../locales/i18n';
import {ETH_SIGN_FRICTION_TEXTFIELD_TEST_ID} from "./EthSignFriction.testIds";

const mockEngine = Engine;

jest.mock('../../../../../core/Engine', () => ({
  init: () => mockEngine.init({}),
}));

jest.mock('react-native-safe-area-context', () => {
  const inset = { top: 0, right: 0, bottom: 0, left: 0 };
  return {
    SafeAreaProvider: jest.fn().mockImplementation(({ children }) => children),
    SafeAreaConsumer: jest
      .fn()
      .mockImplementation(({ children }) => children(inset)),
    useSafeAreaInsets: jest.fn().mockImplementation(() => inset),
  };
});

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
    }),
  };
});

describe('Eth_sign friction bottom sheet', () => {
  it('should render initial friction step correctly', () => {
    const wrapper = shallow(<EthSignFriction />);
    expect(wrapper).toMatchSnapshot();
  });

  it.only('should have checkbox and textfield unlocking steps', async () => {
    const { debug, getAllByText, queryAllByRole, getByRole, getByTestId } = renderWithProvider(
      <EthSignFriction />,
      {
        state: {},
      },
    );

    expect(
      getAllByText(strings('app_settings.toggleEthSignModalCheckBox')).length,
    ).toBe(1);
    expect(
      getAllByText(strings('app_settings.toggleEthSignContinueButton')).length,
    ).toBe(1);

    const continueButton = getByRole('button',
      {name: strings('app_settings.toggleEthSignContinueButton')}
    );

    // Initially disabled continue button
    expect(continueButton.props.disabled).toBeTruthy();

    // Check the checkbox to enable continue button
    fireEvent(getByRole('checkbox'), 'onValueChange', true);
    expect(continueButton.props.disabled).toBeFalsy();

    // press continue button to move to step 2
    fireEvent.press(continueButton);

    // expect the continue button to be disabled again and text changed to 'enable'
    // as the bottom sheet changed to the second step
    const enableButton = getByRole('button',
      {name: strings('app_settings.toggleEthSignEnableButton')}
    );
    expect(enableButton.props.disabled).toBeTruthy();

    // Check the checkbox and continue button are gone
    expect(queryAllByRole('checkbox').length).toBe(0);
    expect(queryAllByRole('button', {name: strings('app_settings.toggleEthSignContinueButton')} ).length).toBe(0);

    debug();

    const textField = getByTestId(ETH_SIGN_FRICTION_TEXTFIELD_TEST_ID);
    expect(enableButton.props.value).toBe('');// initially empty


  });
});
