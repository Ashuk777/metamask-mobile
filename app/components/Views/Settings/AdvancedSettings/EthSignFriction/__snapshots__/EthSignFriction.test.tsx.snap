// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Eth_sign friction bottom sheet should render initial friction step correctly 1`] = `
<ForwardRef>
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "padding": 16,
        "width": "100%",
      }
    }
  >
    <Icon
      color="#D73847"
      name="Danger"
      size="32"
      style={
        Object {
          "alignSelf": "center",
          "color": "#D73847",
          "marginBottom": 4,
          "marginEnd": 3,
          "marginStart": 3,
          "marginTop": 2,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#24272A",
          "fontFamily": "EuclidCircularB-Bold",
          "fontSize": 20,
          "fontWeight": "600",
          "lineHeight": 26,
          "marginHorizontal": 6,
          "textAlign": "center",
        }
      }
    >
      Use at your own risk
    </Text>
    <Text>
      <Text>
        Allowing eth_sign requests can make you vulnerable to phishing attacks. Always review the URL and be careful when signing messages that contain code.
      </Text>
      <Text
        link={true}
        noMargin={true}
        onPress={[Function]}
        primary={true}
      >
         
        Learn more.
      </Text>
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#D7384719",
          "borderLeftColor": "#D73847",
          "borderLeftWidth": 4,
          "borderRadius": 4,
          "color": "#24272A",
          "display": "flex",
          "flexDirection": "row",
          "fontFamily": "EuclidCircularB-Regular",
          "fontSize": 14,
          "fontWeight": "400",
          "lineHeight": 22,
          "marginBottom": 16,
          "marginTop": 24,
          "padding": 8,
        }
      }
    >
      <Icon
        color="#D73847"
        name="Danger"
        size="24"
        style={
          Object {
            "alignSelf": "center",
            "color": "#D73847",
            "marginBottom": 4,
            "marginEnd": 3,
            "marginStart": 3,
            "marginTop": 2,
          }
        }
      />
      <Text
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text>
          If you've been asked to turn this setting on,
        </Text>
        <Text
          style={
            Object {
              "fontFamily": "EuclidCircularB-Bold",
              "fontWeight": "600",
            }
          }
        >
           you might be getting scammed.
        </Text>
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "stretch",
            "display": "flex",
            "flexDirection": "row",
            "fontSize": 14,
            "justifyContent": "flex-start",
            "margin": 16,
          },
        ]
      }
    >
      <Checkbox
        isChecked={false}
        onPress={[Function]}
        style={
          Object {
            "marginEnd": 8,
          }
        }
      />
      <Text
        onPress={[Function]}
        style={
          Object {
            "marginEnd": 16,
          }
        }
      >
        I understand that I can lose all of my funds and NFTs if I enable eth_sign requests.
      </Text>
    </View>
    <View
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
        }
      }
    >
      <Button
        label="Cancel"
        onPress={[Function]}
        style={
          Object {
            "flex": 1,
            "marginEnd": 8,
          }
        }
        variant="Secondary"
        width="full"
      />
      <Button
        disabled={true}
        isDanger={false}
        label="Continue"
        onPress={[Function]}
        style={
          Object {
            "flex": 1,
            "marginStart": 8,
          }
        }
        variant="Primary"
        width="full"
      />
    </View>
  </View>
</ForwardRef>
`;

exports[`Eth_sign friction bottom sheet should render start friction step correctly 1`] = `
<ForwardRef>
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "padding": 16,
        "width": "100%",
      }
    }
  >
    <Icon
      color="#D73847"
      name="Danger"
      size="32"
      style={
        Object {
          "alignSelf": "center",
          "color": "#D73847",
          "marginBottom": 4,
          "marginEnd": 3,
          "marginStart": 3,
          "marginTop": 2,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#24272A",
          "fontFamily": "EuclidCircularB-Bold",
          "fontSize": 20,
          "fontWeight": "600",
          "lineHeight": 26,
          "marginHorizontal": 6,
          "textAlign": "center",
        }
      }
    >
      Use at your own risk
    </Text>
    <Text>
      <Text>
        Allowing eth_sign requests can make you vulnerable to phishing attacks. Always review the URL and be careful when signing messages that contain code.
      </Text>
      <Text
        link={true}
        noMargin={true}
        onPress={[Function]}
        primary={true}
      >
         
        Learn more.
      </Text>
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#D7384719",
          "borderLeftColor": "#D73847",
          "borderLeftWidth": 4,
          "borderRadius": 4,
          "color": "#24272A",
          "display": "flex",
          "flexDirection": "row",
          "fontFamily": "EuclidCircularB-Regular",
          "fontSize": 14,
          "fontWeight": "400",
          "lineHeight": 22,
          "marginBottom": 16,
          "marginTop": 24,
          "padding": 8,
        }
      }
    >
      <Icon
        color="#D73847"
        name="Danger"
        size="24"
        style={
          Object {
            "alignSelf": "center",
            "color": "#D73847",
            "marginBottom": 4,
            "marginEnd": 3,
            "marginStart": 3,
            "marginTop": 2,
          }
        }
      />
      <Text
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text>
          If you've been asked to turn this setting on,
        </Text>
        <Text
          style={
            Object {
              "fontFamily": "EuclidCircularB-Bold",
              "fontWeight": "600",
            }
          }
        >
           you might be getting scammed.
        </Text>
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "stretch",
            "display": "flex",
            "flexDirection": "row",
            "fontSize": 14,
            "justifyContent": "flex-start",
            "margin": 16,
          },
        ]
      }
    >
      <ForwardRef(CheckBoxWithRef)
        boxType="square"
        onValueChange={[Function]}
        style={
          Object {
            "height": 20,
            "marginEnd": 8,
            "width": 20,
          }
        }
        tintColors={
          Object {
            "false": "#BBC0C5",
            "true": "#0376C9",
          }
        }
        value={false}
      />
      <Text
        onPress={[Function]}
        style={
          Object {
            "marginEnd": 16,
          }
        }
      >
        I understand that I can lose all of my funds and NFTs if I enable eth_sign requests.
      </Text>
    </View>
    <View
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
        }
      }
    >
      <StyledButton
        containerStyle={
          Object {
            "flex": 1,
            "marginHorizontal": 8,
          }
        }
        disabledContainerStyle={
          Object {
            "opacity": 0.6,
          }
        }
        onPress={[Function]}
        styleDisabled={
          Object {
            "opacity": 0.6,
          }
        }
        type="cancel"
      >
        Cancel
      </StyledButton>
      <StyledButton
        containerStyle={
          Object {
            "flex": 1,
            "marginHorizontal": 8,
          }
        }
        disabled={true}
        disabledContainerStyle={
          Object {
            "opacity": 0.6,
          }
        }
        onPress={[Function]}
        styleDisabled={
          Object {
            "opacity": 0.6,
          }
        }
        type="confirm"
      >
        Continue
      </StyledButton>
    </View>
  </View>
</ForwardRef>
`;
